### DISCUSSION TITLE
saspy options for logging

### DISCUSSION THREAD

**Discussion by FibonacciTrees** [sassoftware/saspy#534] [Q&A] [answered]
Hi there,

I am trying to use a more efficient way to save a sas log to a specific file. I tried doing this by passing in the log file as an option, but it doesn't work. The code runs, but log not saved to a file. I tried in batch mode and not. I ended up using the method I was already aware of, which is running in batch mode and then writing the ['LOG'] to a file.

What I am trying to do:
`sas = saspy.SASsession(cfgname='default', options=[ "-log", "mylog.log"])
code = open('mycode.sas').read()
res = sas.submit(code)`
        
What I ended up doing:
`logfile = "mylog.log"
sas = saspy.SASsession(cfgname='default')
sas.set_batch(True)
code = open('mycode.sas').read()
    res = sas.submit(code)
    with open(logfile, 'w') as outlog:
        outlog.write(res['LOG'])`

**Comment by tomweber-sas** [comment, marked as answer]
Right, you can't redirect the log out from under saspy, it needs the log to function correctly. SASPy stores the SASLOG for the whole session; all the code you run, not just one submit. You can access the log at any point with the sas.saslog() method, which returns the whole log, up to that point. So If you want to write the log to a file when you're done with your session, just do that:
```
import saspy
sas = saspy.SASsesion()

sas. ...    all the code you want to run

fd = open('out.log','w')
fd.write(sas.saslog())
fd.close()
sas.endsas()
```

**Reply by FibonacciTrees** [reply]
Thank you for confirming! From looking at the source [code](https://github.com/sassoftware/saspy/blob/b3e86ae53601f9731c0ed907c28dc43dd0298fb4/saspy/sasiostdio.py#L971), I take it the workaround I wrote is using code closer to the source. Looks like `sas.saslog()` is getting the same sassession ['LOG'] info I grab.

**Reply by tomweber-sas** [reply]
You were getting the log from running submit(code), and not the entire log. Though, if that's the only thing you ran, all you missed would be the startup part of the log (some boiler plate code I run to get some session info). If you ran other saspy methods, sas.saslog() would have all of that in it too, while the log from the dictionary returned from submit() would only ever have the part of the log from that submit and not everything from the life of the session.

So, you can use whichever you want, depending upon what you're looking to write out. For the code you showed, only doing a single submit of a whole SAS program, what you have is the relevant part of the log. The first bit that's not there, that is in sas.saslog() would just be the same startup stuff, which you may or may not care about.
